### PR TITLE
[GOSSIP][BREAKING] Replace `floodPublish` param with `floodPublishMaxMessageSizeThreshold` 

### LINKED ISSUE #381: Consider converting floodPublish Gossip parameter to message size based
### Description

Introduce `flood_publish_max_size: Long` option instead of `flood_publish: Boolean`
'flood publish' feature would be used just for the messages of size above the specified threshold 
former `flood_publish = false` would be equivalent to `flood_publish_max_size = 0`
former `flood_publish = true` would be equivalent to `flood_publish_max_size = Long.MAX_VALUE`

### Motivation

`floodPublish` ([spec](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md#flood-publishing)) is a good feature to mitigate eclipse attacks against a publisher from one hand. 
From the other hand it may significantly delay message publishing. Especially messages of larger size with limited outbound bandwidth. 
For example let's consider publishing a 1Mb message with bandwidth of 100Mbit/sec. Publishing it without `floodPublish` feature would take roughly **1sec** with mesh size of 10 peers. Publishing it with `floodPublish` feature with 100 peers on the topic would take roughly **10sec**. 
So instead of completely disabling `floodPublish` for limited bandwidth setups it makes sense to apply the feature depending on the message size. So the messages of the size above the threshold would be sent without utilizing `floodPublish`. 
The former 

### Current Implementation

The `floodPublish` parameter is currently `boolean` and is capable of just totally switching on/off flood publish feature 

### Are you planning to do it yourself in a pull request ?

Maybe

### PR DESCRIPTION
**Pull request by tbenr**
fixes #381

### CODE REVIEW

File: libp2p/src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt [hunk 1]
```diff
@@ -400,8 +400,10 @@ open class GossipRouter(
     override fun broadcastOutbound(msg: PubsubMessage): CompletableFuture<Unit> {
         msg.topics.forEach { lastPublished[it] = currentTimeSupplier() }
 
+        val floodPublish = msg.protobufMessage.data.size() <= params.floodPublishMaxMessageSizeThreshold
```
Comment (Nashatyrev) [inline libp2p/src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:403]: We already refer to 'message size' in a couple of places. I would add this notion to `PubsubMessage` interface: 
```patch
Index: libp2p/src/main/kotlin/io/libp2p/pubsub/PubsubRouter.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/libp2p/src/main/kotlin/io/libp2p/pubsub/PubsubRouter.kt b/libp2p/src/main/kotlin/io/libp2p/pubsub/PubsubRouter.kt
--- a/libp2p/src/main/kotlin/io/libp2p/pubsub/PubsubRouter.kt	(revision 1defc232bebd9bb3eaafd05be1e7fa14f436f328)
+++ b/libp2p/src/main/kotlin/io/libp2p/pubsub/PubsubRouter.kt	(date 1729512008659)
@@ -22,6 +22,9 @@
     val topics: List<Topic>
         get() = protobufMessage.topicIDsList
 
+    val size: Long
+        get() = protobufMessage.data.size().toLong()
+
     fun messageSha256() = sha256(protobufMessage.toByteArray())
 
     override fun equals(other: Any?): Boolean
```

**Review by Nashatyrev — APPROVED** [review]
LGTM,
just several nits

**Comment by Nashatyrev** [comment]
Oh, would be great to have a unit test(s) for this parameter

File: libp2p/src/main/kotlin/io/libp2p/pubsub/gossip/GossipParams.kt [hunk 2]
```diff
@@ -112,11 +112,13 @@ data class GossipParams(
...
-     * When this is enabled, published messages are forwarded to all peers with score >=
-     * to publishThreshold
+     * [floodPublishMaxMessageSizeThreshold] controls the maximum size (in bytes) a message will be
+     * published using flood publishing mode.
+     * When a message size is <= [floodPublishMaxMessageSizeThreshold], published messages are forwarded
+     * to all peers with score >= to [GossipScoreParams.publishThreshold]
+     * The default is 0 KiB (never flood publish).
      */
-    val floodPublish: Boolean = false,
+    val floodPublishMaxMessageSizeThreshold: Int = 0,
```
[the remark this answers is not in the document]
Comment (tbenr) [reply libp2p/src/main/kotlin/io/libp2p/pubsub/gossip/GossipParams.kt:121]: yes that makes sense

File: libp2p/src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt (hunk 1 above)
Comment (tbenr) [reply libp2p/src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:403]: right, yes.. seems like you like the idea to expose `Long` even if under the hood we can't go over 2gb (protobuf limit?)

Comment (Nashatyrev) [reply libp2p/src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt:403]: Oh, it may be an overkill. Let's stick with just Int then

**Pull request merged**